### ISSUE TITLE
[Python] Standardize exception handling

### ISSUE THREAD

**Issue by yuanming-hu** [taichi-dev/taichi#1066]
Two nice tips raised by @rexwangcc:

> * It seems we've been using `assert` over `raise Exception` acorss the codebase, while they works similar in most cases, `assert` statements are removed when the Py compilation is optimized (e.g. python -O examples/***.py), and thus assertions will be unsafely bypassed. Do we have special considerations about why we chose `assert`s?
> * For the places that are raising exceptions, I noticed often wild `Exception` is being thrown, which might not be a good practice to me. As we start to discuss refactor the operations and math libs, it might be a good chance to have more explicit Exceptions defined in `exception.py`?

There seems to be a few things to decide:
 - Do we abandon raw `assert`s in non-testing code?
 - Shall we avoid throwing raw `Exception`s, so that users have clearer error messages?
 - If we enforce the use of derived classes of `Exception` such as `ValueError`, do we need to define a set of customized exceptions such as `TaichiValueError` in  https://github.com/taichi-dev/taichi/blob/master/python/taichi/lang/exception.py?

**Comment by k-ye** [comment]
I think the distinction between assertions and exceptions are: whether this is a violation of the system invariants, or an error possibly due to something we cannot control (e.g. user input error, OOM, etc). In another word, assertions are something that should never fail. Otherwise we should throw an exception.

**Comment by liaopeiyuan** [comment]
Agreed. I may be of help in this, as I'm currently refactoring the matrix and math api, where I've found some of the error messages (resulted from incorrect usage) to be really cryptic.....
